### PR TITLE
Implement Real Time Transfer (RTT pipe) device driver

### PR DESCRIPTION
**Pull request by gerard5**
## Description

`Real Time Transfer`, a.k.a. `RTT` is an interface specified by SEGGER and implemented in their tools and freely implemented in OpenOCD. It is based on basic memory reads and writes to transfer data bidirectionally between target and host. The specification is independent of the target architecture. Every target that supports so called _"background memory access"_, which means that the target memory can be accessed by the debugger while the target is running, can be used. This interface is the fastest alternative for semihosting. It's speed is determined by the debug probe speed, (this implementation was tested using 12MHz and 24MHz JTAG probe; 6MHz and 4MHz SWD).

This change implements two RTT channels:
1) `console` which is used `in parallel to UART console` (but through debug interface)
2) `auxiliary pipe` for `phfs` protocol used by `phoenixd` server

To use with OpenOCD, the setup is:
```tcltk
rtt setup 0x20057f00 256 "RTT PIPE"
rtt server start 18021 0
rtt server start 18022 1
rtt polling_interval 20
```

Which initializes two TCP servers on `18021` and `18022` ports which corresponds to two channels described above (`console` and `phoenixd`).

*NOTE: Given in example `0x2003ff00` address (for rt1176) is set using macro define `RTT_ADDR` in `peripheral.h`, and might be overridden in per project `board_config.h`, the address provides location of RTT descriptors (last 256 bytes of DTCM)
Default RTT tag `"RTT PIPE"` may be overridden using `RTT_TAG_BACKWARD` define (written backwards).

To connect phoenixd run (imxrt117x is an example target):
```
phoenixd -s _boot/armv7m7-imxrt117x-evk/ -t 127.0.0.1:18022
```

Depending on configuration of OpenOCD, any IP address can be specified and remote programming and console access is possible.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `imxrt1176-nil`, `imxrt1064-evk`, `stm32l4a6-nucleo`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: devices/pipe-rtt/device.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ */
+
+#include <devices/devs.h>
+#include <hal/hal.h>
+#include <lib/lib.h>
+
+
+extern int rtt_check(int chan);
+extern ssize_t rtt_write(int chan, const void *buf, size_t count);
+extern ssize_t rtt_read(int chan, void *buf, size_t count);
```
Comment (agkaminski) [inline devices/pipe-rtt/device.c:23]: Why not provide these prototypes in a header file?

File: devices/pipe-rtt/rtt.c [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+	rtt->txChannel[chan].wr = wr;
+
+	hal_cpuDataMemoryBarrier();
+
+	return count - todo;
+}
+
+
+ssize_t rtt_writeBlocking(int chan, const void *buf, size_t count)
```
Comment (agkaminski) [inline devices/pipe-rtt/rtt.c:136]:
```suggestion
static ssize_t rtt_writeBlocking(int chan, const void *buf, size_t count)
```

**Review by agkaminski — COMMENTED** [review]
Small nitpicks, I really like `*Hook` approach :)

File: devices/pipe-rtt/rtt.c (hunk 2 above)
Comment (gerard5) [reply devices/pipe-rtt/rtt.c]: Done.

File: devices/pipe-rtt/device.c (hunk 1 above)
Comment (gerard5) [reply devices/pipe-rtt/device.c]: Done.

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**